internal/controller/githubissue_controller.go

```diff
@@ -0,0 +1,143 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// GithubIssueReconciler reconciles a GithubIssue object
+type GithubIssueReconciler struct {
+	client.Client
+	Scheme      *runtime.Scheme
+	Log         *zap.Logger
+	IssueClient IssueClient
+	Recorder    record.EventRecorder
+}
+
+// +kubebuilder:rbac:groups=issues.dana.io,resources=githubissues,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=issues.dana.io,resources=githubissues/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=issues.dana.io,resources=githubissues/finalizers,verbs=update
+// +kubebuilder:rbac:groups="",resources=events,verbs=create;watch;list
+
+func (r *GithubIssueReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log
+
+	var issueObject = &issuesv1alpha1.GithubIssue{}
+	if err := r.Get(ctx, req.NamespacedName, issueObject); err != nil {
+		if client.IgnoreNotFound(err) != nil {
+			log.Error("unable to fetch issue object", zap.Error(err))
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	splitUrl := strings.Split(issueObject.Spec.Repo, "/")
+	owner := splitUrl[3]
+	repo := splitUrl[4]
+
+	log.Info(fmt.Sprintf("attempting to get issues from %s/%s", owner, repo))
+	gitHubIssue, err := r.FindIssue(ctx, owner, repo, issueObject)
+	if err != nil {
+		log.Error("failed fetching issue", zap.Error(err))
+		return ctrl.Result{}, err
+	}
+
+	if !issueObject.ObjectMeta.DeletionTimestamp.IsZero() {
+		log.Info("closing issue")
+
+		if gitHubIssue == nil {
+			return ctrl.Result{}, fmt.Errorf("cannot close issue: gitHubIssue is nil")
+		}
+
+		if err := r.CloseIssue(ctx, owner, repo, gitHubIssue); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed closing issue: %v", err)
+		}
+		if err := finalizer.Cleanup(ctx, r.Client, issueObject, r.Log); err != nil {
+			log.Error("failed cleaning up finalizer", zap.Error(err))
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+	err = finalizer.Ensure(ctx, r.Client, issueObject, r.Log)
+	if err != nil {
+		log.Error("failed adding finalizer", zap.Error(err))
+		return ctrl.Result{}, err
+	}
+
+	if gitHubIssue == nil {
+		log.Info("creating issue")
+		if err := r.CreateIssue(ctx, owner, repo, issueObject); err != nil {
+			log.Error("failed creating issue", zap.Error(err))
+			return ctrl.Result{}, err
+		}
+
+		gitHubIssue, err = r.FindIssue(ctx, owner, repo, issueObject)
+		if err != nil {
+			log.Error("failed fetching newly created issue", zap.Error(err))
+			return ctrl.Result{}, err
+		}
+		if gitHubIssue == nil {
+			log.Warn("Cannot update status: githubIssue is nil", zap.String("IssueName", issueObject.Name), zap.String("Namespace", issueObject.Namespace))
+		} else {
+			if err := r.UpdateIssueStatus(ctx, issueObject, gitHubIssue); err != nil {
+				log.Error("failed updating issue status", zap.Error(err))
+			}
+		}
+		log.Info("issue created successfully")
+		return ctrl.Result{}, nil
```

Comment:
I'd move most of this logic to a separate function, like `handleNewIssue` or something like this